RFC-0009-jdbc-join-push-down.md

```diff
@@ -0,0 +1,570 @@
+The changes help in finding what all can be pushed down and able to be translated. The filters that can't be pushed down are kept in a new FilterNode.
+
+## Changes in SPI module 
+
+- Added 2 new classes : JoinTableInfo, JoinTableSet
+- JoinTableInfo holds the information about a table that is undergoing pushdown. It has details about tableHandle, assignments and outputVariables.
+- JoinTableSet is used to store a set of JoinTableInfo objects.
+- These are explained in detail below.
+
+We have added a new Connector capability called SUPPORTS_JOIN_PUSHDOWN ConnectorCapabilities enum in SPI. Any connector which wants to participate in pushdown need to added this Connector capability to their Set of ConnectorCapabilities. JDBC has done it like [this](#3-use-multijoinnode-to-group-jdbc-tables-based-on-connector-name)
+
+## Edge cases
+
+### Self Join
+Self join is a scenario where a table is joined with itself. This works natively in Presto. When you do a self join in presto, that is sent as select statements to the datasource first. On top of that presto does all the joining and filtering etc. 
+We want to push down this self join to database level. This is similar to pushing down an inner join. But, we faced some difficulties in achieving this.
+
+#### Difficulty faced :
+If we do a join query which joins the same table with itself, We get 2 TableScans in the plan.
+There isn’t a way to differentiate between the columns of these TableScans.
+
+Example : Assume there is a column 'col_1' in a table 'table_1'.  We join table_1 with itself ->
+```
+select * from table_1 a join table_1 b on a.col_1 = b.col_1;
+```
+The column 'col_1' will be referred as col_1 (in outputVariables) inside TableScan1 and will point to 'col_1' JdbcColumnHandle (in assignments)
+It will be referred as col_1_0 (in outputVariables) inside TableScan2 and will point to 'col_1' JdbcColumnHandle (in assignments)
+When our Optimiser 'JdbcJoinRenderByConnector' is run, both these TableScans get converted to a Single TableScan, with assignments, outputVariables, etc. combined.
+(Note : assignments now has two references to the same column 'col_1'. One from 'col_1' and the other from 'col_1_0')
+This creates a problem in pushPredicateIntoTableScan method in PickTableLayout class, which assumes that all the values in assignments are unique.
+This assumption by presto is done on the basis that till now all queries done by presto to any database is as a select from a single table.
+
+#### Proposed Solution :
+
+We can bring in a 'table alias' to uniquely identify the tables. This is not the table alias that the user passes in his query.
+This alias can then be used to unique identify columns and assignments even in the case of self joins.
+This alias can then also be used in the Join conditions, predicates, filters, etc.. which will simplify the Pushdown Query that we create in Query Builder.
+This solved the issue and we were able to push down self joins as well.
```

Comment:
```suggestion
This alias can then be used to build the pushdown filter clauses from the query predicates
```
`conditions, predicates, filters,` all refer to the same thing, so lets make this more precise